### PR TITLE
refactor: remove more compiler runtime code

### PR DESCRIPTION
**Pull request by bartlomieju**
I'm going all in on this one - rewriting some stuff to trim down the runtime - eg. use `Deno.core.print()` for debug logs instead of console.

### CODE REVIEW

**Comment by ry** [comment]
Tip: run "cargo test bundle" often ... it seems to be a edge case that breaks easily.

File: cli/ops/compiler.rs [hunk 1]
```diff
@@ -1,13 +1,59 @@
...
+  }
+}
+
+fn op_compiler_respond(
+  response: Arc<Mutex<Option<String>>>,
+  args: Value,
+  _zero_copy: &mut [ZeroCopyBuf],
+) -> Result<JsonOp, OpError> {
+  let mut r = response.lock().unwrap();
+  *r = Some(args.to_string());
```
Comment (ry) [inline cli/ops/compiler.rs:57]: What's happening here? `response` is coming from `init()` ? Since `init()` is only called once per process... does that mean that the response is only possible to be set once? What if this is called twice?

File: cli/tsc/99_main_compiler.js [hunk 2]
```diff
@@ -1804,64 +1827,83 @@ delete Object.prototype.__proto__;
...
+  }
+
+  let hasBootstrapped = false;
+
+  function bootstrapCompilerRuntime() {
+    if (hasBootstrapped) {
+      throw new Error("Worker runtime already bootstrapped");
+    }
+    hasBootstrapped = true;
+    globalThis.__bootstrap = undefined;
```
Comment (ry) [inline cli/tsc/99_main_compiler.js:1903]: Side note 1: we should combine the `__bootstrap` and `bootstrap` global variables.

Side note 2: cli/js2/base64.js doesn't use __bootstrap it uses __base64

File: cli/tsc/99_main_compiler.js [hunk 3]
```diff
@@ -1804,64 +1827,83 @@ delete Object.prototype.__proto__;
...
+    runtimeStart("TS");
+  }
+
+  globalThis.bootstrapCompilerRuntime = bootstrapCompilerRuntime;
+  globalThis.tsCompilerOnMessage = tsCompilerOnMessage;
```
Comment (ry) [inline cli/tsc/99_main_compiler.js:1908]: much simpler

File: cli/ops/compiler.rs (hunk 1 above)
Comment (bartlomieju) [reply cli/ops/compiler.rs:57]: > What's happening here? `response` is coming from `init()` ? Since `init()` is only called once per process... 

That's just op registration, this op is used after TS compiler does it's job and returns the result. Before this change it was happening using `globalThis.postMessage()` using Web Worker APIs.

> does that mean that the response is only possible to be set once? What if this is called twice?

That's correct, and that has been the behavior for quite some time. TS compiler is a single-use thing, ie. only single request can be sent and then it must be discarded.

File: cli/tsc/99_main_compiler.js (hunk 2 above)
Comment (bartlomieju) [reply cli/tsc/99_main_compiler.js:1903]: > Side note 1: we should combine the `__bootstrap` and `bootstrap` global variables.
 
`bootstrap` is not used in `cli/tsc/`, if you talk about `cli/js2/` then that's good I guess 👍 

> Side note 2: cli/js2/base64.js doesn't use __bootstrap it uses __base64

Noted

File: cli/ops/compiler.rs (hunk 1 above)
Comment (ry) [reply cli/ops/compiler.rs:57]: Maybe you should add `assert!(r.is_none());` before this

Comment (bartlomieju) [reply cli/ops/compiler.rs:57]: Done

**Comment by bartlomieju** [comment]
> What is the size of the `COMPILER_SNAPSHOT.bin` now?

this PR:

```
"COMPILER_SNAPSHOT.bin": 3049099
```

`master`:

```
"COMPILER_SNAPSHOT.bin": 3102191
```

50kB saved 🎉 🤣

**Comment by ry** [comment]
It’s mentally simpler tho :)

I just want to see that the size is indeed decreasing
